python/llm/src/ipex_llm/transformers/pipeline_parallel.py

```diff
@@ -307,3 +311,304 @@ def pipeline_parallel_generate(self,
+    def __init__(self, checkpoint, rank, world_size, low_bit, max_num_seqs,
+                 torch_dtype=torch.float16):
+        self.pp_config = PPConfig(rank, world_size)
+        self.dtype = torch_dtype
+        start = time.perf_counter()
+        model = self.load_model(checkpoint, world_size, low_bit)
+        end = time.perf_counter()
+        logger.info(f"Time to load weights: {end - start:.2f}s")
+
+        self.model = model
+        self.rank = rank
+        self.world_size = world_size
+        self.pre_rank = (self.rank - 1) % self.world_size
+        self.next_rank = (self.rank + 1) % self.world_size
+        self.hidden_size = self.model.config.hidden_size
+        self.max_num_seqs = max_num_seqs
+        self.on_going_batches = [None] * self.world_size
+        self.input_ids_dict = {}
+        self.past_key_values_dict = {}
+        self.tokens = {}
+        self.token_times = {}
+        self.waiting_requests = asyncio.Queue()
+        self.send_buff = None
+        self.dict_lock = threading.Lock()
+        self.streamer = {}
+        self.token_cache = {}
+        self.print_len = {}
+        self.is_finish = {}
+        self.model_name = checkpoint
+        self.layer_start = 0
+
+    def load_model(self, model_path, world_size, low_bit='sym_int4'):
+        from ipex_llm.transformers import AutoModelForCausalLM, AutoModel
+        try:
+            model = AutoModelForCausalLM.from_pretrained(model_path,
+                                                         load_in_low_bit=low_bit,
+                                                         torch_dtype=self.dtype,
+                                                         optimize_model=True,
+                                                         trust_remote_code=True,
+                                                         use_cache=True,
+                                                         pipeline_parallel_stages=world_size)
+        except:
+            model = AutoModel.from_pretrained(model_path,
+                                              load_in_low_bit=low_bit,
+                                              torch_dtype=self.dtype,
+                                              optimize_model=True,
+                                              trust_remote_code=True,
+                                              use_cache=True,
+                                              pipeline_parallel_stages=world_size)
+        model = model.eval()
+        return model
+
+    def model_step(self, input, cur_batch):
+        if cur_batch is None or cur_batch.stopped or input is None:
+            return None
+
+        cur_id = cur_batch.batch_id
+        _past_key_values = self.past_key_values_dict.get(cur_id, None)
+
+        if self.rank == 0:
+            input_ids = input
+            inputs_embeds = None
+        else:
+            input_ids = None
+            inputs_embeds = input
+
+        output = self.model(input_ids=input_ids,
```

Comment:
Need to pass `attention_mask = make_attention_mask(cur_batch.prompt_lengths).to(input.device)` here for batch serving.

Reply:
> Need to pass `attention_mask = make_attention_mask(cur_batch.prompt_lengths).to(input.device)` here for batch serving.

Have updated :)